workitem/workitem_repository.go

```diff
@@ -116,6 +181,185 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, workitemID string,
+	Item := WorkItem{}
+	var tx *gorm.DB
+	switch secondItemDirection {
+	case DirectionBelow:
+		// Finds the item below which reorder item has to be placed
+		tx = r.db.Where("execution_order < ?", order).Order("execution_order desc", true).Last(&Item)
+	case DirectionAbove:
+		// Finds the item above which reorder item has to be placed
+		tx = r.db.Where("execution_order > ?", order).Order("execution_order", true).Last(&Item)
+	default:
+		return nil, nil, nil
+	}
+	if tx.RecordNotFound() {
+		return nil, nil, tx.Error
+	}
+	if tx.Error != nil {
+		return nil, nil, errors.NewInternalError(tx.Error.Error())
+	}
+
+	ItemId := strconv.FormatUint(Item.ID, 10)
+	return &ItemId, &Item.ExecutionOrder, nil
+
+}
+
+// FindFirstItem returns the order of the target workitem
+func (r *GormWorkItemRepository) FindFirstItem(id string) (*float64, error) {
+	Item := WorkItem{}
+	Id, err := strconv.ParseUint(id, 10, 64)
+	if err != nil || Id == 0 {
+		return nil, errors.NewNotFoundError("work item", string(Id))
+	}
+	tx := r.db.First(&Item, Id)
+	if tx.RecordNotFound() {
+		return nil, errors.NewNotFoundError("work item", id)
+	}
+	if tx.Error != nil {
+		return nil, errors.NewInternalError(tx.Error.Error())
+	}
+	return &Item.ExecutionOrder, nil
+}
+
+// Reorder places the to-be-reordered workitem above the input workitem.
+// The order of workitems are spaced by a factor of 1000.
+// The new order of workitem := (order of previousitem + order of nextitem)/2
+// Version must be the same as the one int the stored version
+func (r *GormWorkItemRepository) Reorder(ctx context.Context, direction DirectionType, targetID *string, wi app.WorkItem, modifierID uuid.UUID) (*app.WorkItem, error) {
+	var order float64
+	res := WorkItem{}
+
+	id, err := strconv.ParseUint(wi.ID, 10, 64)
+	if err != nil || id == 0 {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+
+	tx := r.db.First(&res, id)
+	if tx.RecordNotFound() {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+	if err := tx.Error; err != nil {
+		return nil, errors.NewInternalError(err.Error())
+	}
+	if res.Version != wi.Version {
+		return nil, errors.NewVersionConflictError("version conflict")
+	}
+
+	wiType, err := r.witr.LoadTypeFromDB(ctx, wi.Type)
+	if err != nil {
+		return nil, errors.NewBadParameterError("Type", wi.Type)
+	}
+
+	switch direction {
+	case DirectionAbove:
+		// if direction == "above", place the reorder item **above** the workitem having id equal to targetID
+		aboveItemOrder, err := r.FindFirstItem(*targetID)
+		if aboveItemOrder == nil || err != nil {
+			return nil, errors.NewNotFoundError("work item", *targetID)
+		}
+		belowItemId, belowItemOrder, err := r.FindSecondItem(aboveItemOrder, DirectionBelow, wi.ID)
+		if err != nil {
+			return nil, errors.NewNotFoundError("work item", *targetID)
+		}
+		if *belowItemId == "0" {
```

Comment:
There is no test case that test the two special cases where you want to move something 'before' the first, or 'after' the last wi? And you can't determine the first being the wi with id 0